src/v/archival/ntp_archiver_service.cc

```diff
@@ -415,8 +444,24 @@ ss::future<ntp_archiver::scheduled_upload> ntp_archiver::schedule_single_upload(
     start_upload_offset = offset + model::offset(1);
     auto delta
       = base - _partition->get_offset_translator_state()->from_log_offset(base);
+    // The upload is successful only if both segment and tx_range are uploaded.
+    auto upl_fut
+      = ss::when_all(upload_segment(upload), upload_tx(upload))
+          .then([](auto tup) {
+              auto [fs, ftx] = std::move(tup);
+              auto rs = fs.get();
+              auto rtx = ftx.get();
+              if (
+                rs == cloud_storage::upload_result::success
+                && rtx == cloud_storage::upload_result::success) {
+                  return rs;
+              } else if (rs != cloud_storage::upload_result::success) {
+                  return rs;
+              }
+              return rtx;
```

Comment:
here if not both succeeded, we need to choose which of two errors return. I think we need here some logic base on the values of both errors

Reply:
> if both operations are failed then it will return error from segment upload operation

why?